cmd/gen_token.go

```diff
@@ -0,0 +1,196 @@
+			}
+			fmt.Fprintf(debugOutput(), "GCP ID token fetched is: %s\n", idToken)
+			tokens := [][]byte{[]byte(idToken)}
+			return tokens, nil
+		}
+
+		fmt.Fprintf(debugOutput(), "Attestation Address is set to %s\n", asAddress)
+
+		Region, err := getRegion(mdsClient)
+		if err != nil {
+			return fmt.Errorf("failed to fetch Region from MDS: %v", err)
+		}
+
+		ProjectID, err := mdsClient.ProjectID()
+		if err != nil {
+			return fmt.Errorf("failed to retrieve ProjectID from MDS: %v", err)
+		}
+
+		verifierClient, err := getRESTClient(ctx, asAddress, ProjectID, Region)
+		if err != nil {
+			return fmt.Errorf("failed to create REST verifier client: %v", err)
+		}
+
+		//  supports GCE VM. Hard code the AK type.
+		key = "gceAK"
+		fmt.Fprintf(debugOutput(), "key is set to gceAK\n")
+
+		// Set GCE AK (EK signing) cert
+		if key == "gceAK" {
+			var gceAK *client.Key
+			var err error
+			if keyAlgo == tpm2.AlgRSA {
+				gceAK, err = client.GceAttestationKeyRSA(rwc)
+			}
+			if keyAlgo == tpm2.AlgECC {
+				gceAK, err = client.GceAttestationKeyECC(rwc)
+			}
+			if err != nil {
+				return err
+			}
+			if gceAK.Cert() == nil {
+				return errors.New("failed to find gceAKCert on this VM: try creating a new VM or contacting support")
+			}
+			gceAK.Close()
+		}
+
+		attestAgent := agent.CreateAttestationAgent(rwc, attestationKeys[key][keyAlgo], verifierClient, principalFetcher)
+
+		fmt.Fprintf(debugOutput(), "Fetching attestation verifier OIDC token\n")
+		token, err := attestAgent.Attest(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve attestation service token: %v", err)
+		}
+
+		// Get token expiration.
+		claims := &jwt.RegisteredClaims{}
+		_, _, err = jwt.NewParser().ParseUnverified(string(token), claims)
+		if err != nil {
+			return fmt.Errorf("failed to parse token: %w", err)
+		}
+
+		now := time.Now()
+		if !now.Before(claims.ExpiresAt.Time) {
+			return errors.New("token is expired")
+		}
+
+		// Print out the claims in the jwt payload
+		mapClaims := jwt.MapClaims{}
+		_, _, err = jwt.NewParser().ParseUnverified(string(token), mapClaims)
+		if err != nil {
+			return fmt.Errorf("failed to parse token: %w", err)
+		}
+		claimsString, err := json.MarshalIndent(mapClaims, "", "  ")
+		if err != nil {
+			return fmt.Errorf("failed to format claims: %w", err)
+		}
+
+		fmt.Fprintf(debugOutput(), string(claimsString)+"\n")
+
+		if output == "" {
+			fmt.Fprintf(messageOutput(), string(token)+"\n")
+		}
+
+		if output != "" {
+			out := []byte(token)
+			if _, err := dataOutput().Write(out); err != nil {
+				return fmt.Errorf("failed to write attestation report: %v", err)
+			}
+		}
+
+		return nil
+	},
+}
+
+// getRESTClient returns a REST verifier.Client that points to the given address.
+// It defaults to the Attestation Verifier instance at
+// https://confidentialcomputing.googleapis.com.
+func getRESTClient(ctx context.Context, asAddr string, ProjectID string, Region string) (verifier.Client, error) {
```

Comment:
This is some repetitive code from "go-tpm-tools/launcher/container_runner.go"

Ideally we need to refactor the code to avoid this, but for now at least add some comment or TODO to make them point to each other. (Same for the principalFetcher)